### ISSUE TITLE
Using the trained model

### ISSUE THREAD

**Issue by Ate25436** [LuciusLan/BSL_Negation#2]
Hello,

Thanks to your guidance(#1 ), I was able to successfully train the model using your code.
However, I'm not sure how to perform inference with the trained model.

Could you please provide some instructions or an example on how to use the trained model for inference?

I would really appreciate your help.

**Comment by LuciusLan** [comment]
A minimum code snippet would be as following:
```
# suppose you have saved your trained model and load it with model = torch.load('trained_model_file')
model.eval()
test_sent ='This is a not a test sentence.'
input_ids, scopes, subword_mask, padding_mask = do_process(test_sent) 
# there is no actual function called do_process in the codes. see explanation below.

scope_logits, start_logits, end_logits = model(input_ids, padding_mask)
scope_pred, scope_tar = pack_subword_pred(scope_logits.detach().cpu(), scopes.detach().cpu(), subword_mask.detach().cpu(), padding_mask.cpu())
```

For inference on new text, the `scopes` (ground-truth labels) would be unavailable, it can be replaced with a one tensor same shape as `input_ids`. For `subword_mask` and `padding_mask`  the original processor for the 4 datasets tested is too complicated (and I can't understand them any more...), I'm afraid you have to write your own `do_process` function. 

Basically for `subword_mask` you can loop over the tokenized sentence and check for the special prefix for a continued subword / start of new word:
```
input_ids = tokenizer(test_sent)['input_ids']
input_subwords = tokenizer.convert_ids_to_tokens(input_ids)
# then loop over input_subwords, check for the prefix for start of new word and mark it 1, continued subword as 0. you can find the prefix by looking at the actual decoded input_subwords.
```
`padding_mask` indicate the length of input sentence tokens, padded with 0. you can run a 0 padding over `tokenizer(test_sent)['attention_mask']`


I'm really sorry for the inconvienced caused due to poor maintenance of the project

**Comment by Ate25436** [comment]
Thank you for your quick response.
I'll try it.

**Comment by Ate25436** [comment]
Thanks to your advice, I was able to align the arguments for ```pack_subword_pred```. However, I encountered an error inside the ```pack_subword_pred``` function.

It seems that the cause of the error is the following part of the code (line 614 of util.py), where an inhomogeneous numpy array is being passed:
```
for i, j in zip(logit, mask):
    if j == 1:
        if in_split == 1:
            word_pieces.insert(0, head)
            print(word_pieces)
        ->  mode_pred = np.argmax(np.average(word_pieces, axis=0), axis=0)
            if len(new_logits) > 0:
                new_logits[-1] = mode_pred
            else:
                new_logits.append(mode_pred)
            word_pieces = []
            in_split = 0
        new_logits.append(np.argmax(i))
        head = i
```
Could you please advise on how to resolve this issue?
Sorry for asking so many questions. I'd really appreciate it if you could take a look whenever you have time.
